engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/serialization/KafkaAvroSerializationSchemaFactory.scala

```diff
@@ -1,46 +1,45 @@
 package pl.touk.nussknacker.engine.avro.serialization
 
 import java.lang
+import java.nio.charset.StandardCharsets
 
 import org.apache.avro.Schema
 import org.apache.flink.formats.avro.typeutils.NkSerializableAvroSchema
 import org.apache.flink.streaming.connectors.kafka.KafkaSerializationSchema
 import org.apache.kafka.clients.producer.ProducerRecord
 import org.apache.kafka.common.serialization.Serializer
+import pl.touk.nussknacker.engine.flink.util.keyed.StringKeyedValue
 import pl.touk.nussknacker.engine.kafka.KafkaConfig
-import pl.touk.nussknacker.engine.kafka.serialization.{KafkaProducerHelper, UUIDSerializer}
+import pl.touk.nussknacker.engine.kafka.serialization.KafkaProducerHelper
 
 /**
   * Factory class for Flink's KeyedSerializationSchema. It is extracted for purpose when for creation
-  * of KeyedSerializationSchema are needed additional avro related information.
+  * of KafkaSerializationSchema are needed additional avro related information. SerializationSchema will take
+ *  KafkaSerializationSchema with key extracted in the step before serialization
   */
 trait KafkaAvroSerializationSchemaFactory extends Serializable {
 
-  def create(topic: String, version: Option[Int], schemaOpt: Option[NkSerializableAvroSchema], kafkaConfig: KafkaConfig): KafkaSerializationSchema[AnyRef]
+  def create(topic: String, version: Option[Int], schemaOpt: Option[NkSerializableAvroSchema], kafkaConfig: KafkaConfig): KafkaSerializationSchema[StringKeyedValue[AnyRef]]
 
 }
 
 /**
   * Abstract base implementation of [[pl.touk.nussknacker.engine.kafka.serialization.KafkaSerializationSchemaFactory]]
-  * which uses Kafka's Serializer in returned Flink's KeyedSerializationSchema. It serializes only value - key will be
-  * randomly generated as a UUID.
+  * which uses Kafka's Serializer in returned Flink's KafkaSerializationSchema for value - key will be taken from
+  * step before serialization
   */
 abstract class KafkaAvroValueSerializationSchemaFactory extends KafkaAvroSerializationSchemaFactory {
 
-  protected def createKeySerializer(kafkaConfig: KafkaConfig): Serializer[AnyRef] =
-    new UUIDSerializer[AnyRef]
-
   protected def createValueSerializer(schemaOpt: Option[Schema], version: Option[Int], kafkaConfig: KafkaConfig): Serializer[AnyRef]
 
-  override def create(topic: String, version: Option[Int], schemaOpt: Option[NkSerializableAvroSchema], kafkaConfig: KafkaConfig): KafkaSerializationSchema[AnyRef] = {
-    new KafkaSerializationSchema[AnyRef] {
-      private lazy val keySerializer = createKeySerializer(kafkaConfig)
+  override def create(topic: String, version: Option[Int], schemaOpt: Option[NkSerializableAvroSchema], kafkaConfig: KafkaConfig): KafkaSerializationSchema[StringKeyedValue[AnyRef]] = {
```

Comment:
I think this will make it difficult to add other key serializations without breaking API. Maybe we can move information about StringKeyedValue a bit higher?